src/peft/tuners/xlora/insertion.py

```diff
@@ -0,0 +1,284 @@
+            result = self.target.base_layer(x, *args, **kwargs)
+        elif self.target.merged:
+            result = self.target.base_layer(x, *args, **kwargs)
+        else:
+            result = self.target.base_layer(x, *args, **kwargs)
+            for adapter_n, active_adapter in enumerate(self.target.active_adapters):
+                if active_adapter not in self.target.lora_embedding_A:
+                    continue
+                embedding_A = self.target.lora_embedding_A[active_adapter].T
+                embedding_B = self.target.lora_embedding_B[active_adapter].T
+                scaling = self.target.scaling[active_adapter]
+                x_mod = self.apply_scalings_to_x(x, xlora_scalings, adapter_n)
+                after_A = self.target._embed(x_mod, embedding_A)  # type: ignore
+                result += (after_A @ embedding_B) * scaling * self.config.global_scaling_weight
+
+        return result
+
+
+class xLoRAConv2dLayer(xLoRALayer):
+    def __init__(
+        self,
+        model: nn.Module,  # PeftModel
+        target: lora.Conv2d,
+        target_forward: Callable[..., Any],
+        layer_number: int,
+        config: xLoRAConfig,
+    ) -> None:
+        super().__init__(model, target, target_forward, layer_number, config)
+
+    def forward(self, x: Tensor, *args: Any, **kwargs: Any) -> Tensor:
+        """
+        This method is designed to be a drop-in-replacement for the peft LoRA layers' .forward method.
+        To use it, a bound method must be created (bound to an instance of the xLoRALayer class).
+        """
+
+        previous_dtype = x.dtype
+        xlora_scalings = self.get_maybe_topk_scalings()
+
+        if self.target.disable_adapters:
+            if self.target.merged:
+                self.target.unmerge()
+            result = self.target.base_layer(x, *args, **kwargs)
+        elif self.target.merged:
+            result = self.target.base_layer(x, *args, **kwargs)
+        else:
+            result = self.target.base_layer(x, *args, **kwargs)
+            for adapter_n, active_adapter in enumerate(self.target.active_adapters):
+                if active_adapter not in self.target.lora_A.keys():
+                    continue
+                lora_A = self.target.lora_A[active_adapter]
+                lora_B = self.target.lora_B[active_adapter]
+                dropout = self.target.lora_dropout[active_adapter]
+                scaling = self.target.scaling[active_adapter]
+                x = x.to(lora_A.weight.dtype)  # type: ignore
+                x_mod = self.apply_scalings_to_x(x, xlora_scalings, adapter_n)
+                result += lora_B(lora_A(dropout(x_mod))) * scaling * self.config.global_scaling_weight
+
+        result = result.to(previous_dtype)
+        return result
+
+
+class BaseTunerWrapper:
+    def __init__(self, base_model: BaseTuner, classifier: xLoRAClassifier):
+        self.model = base_model.model
+        self.classifier = classifier
```

Comment:
Why does this need the classifier?

Reply:
It actually does not, thank you for noticing that!